### PR TITLE
Add level and pos context to `PlayerEvent.HarvestCheck`

### PR DESCRIPTION
**Pull request by IchHabeHunger54**
Adds the `BlockToolCheckEvent` to allow mods to check whether a tool is valid for a given block. The event has the usual block event context (level, pos, state), as well as itemstack context and optional entity context. This event is designed to be a replacement for `PlayerEvent.HarvestCheck`, which is currently called immediately afterwards for compatibility reasons.

This was originally suggested to me by @Thodor12 from the Minecolonies team, who requested this functionality for block entity dependent tool requirements. I also see uses for this in some contexts for WAILA derivatives.

Current users of `PlayerEvent.HarvestCheck` should replace this event with `BlockToolCheckEvent`, and a `event.getEntity() instanceof Player` check if needed.

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/common/CommonHooks.java [hunk 1]
```diff
@@ -208,7 +211,13 @@ public static boolean canContinueUsing(@NotNull ItemStack from, @NotNull ItemSta
         return false;
     }
 
-    public static boolean isCorrectToolForDrops(@NotNull BlockState state, @NotNull Player player) {
+    public static boolean isCorrectToolForDrops(@NotNull BlockGetter level, @NotNull BlockPos pos, @NotNull BlockState state, @NotNull Player player) {
+        if (level instanceof LevelAccessor accessor) { //BlockEvent expects a LevelAccessor
+            BlockEvent.ToolCheck event = EventHooks.doToolCheck(accessor, pos, state, player, player.getMainHandItem());
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/CommonHooks.java]:
```suggestion
            BlockEvent.ToolCheck event = NeoForge.EVENT_BUS.post(new BlockEntity.ToolCheck(accessor, pos, state, player, player.getMainHandItem()));
```
if you do this you can also remove the `EventHooks` method.

Comment (IchHabeHunger54) [reply src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: I originally did it this way, then saw the pattern in EventHooks and adjusted to that. Will adjust back, one moment.

Comment (IchHabeHunger54) [reply src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: Adjusted!

File: src/main/java/net/neoforged/neoforge/event/level/BlockEvent.java [hunk 2]
```diff
@@ -464,4 +465,49 @@ public BlockState getFinalState() {
...
+     * Allows for force-allowing or force-denying a tool to break a block,
+     * and for additional checks depending on additional context for the block (e.g. block entity context).
+     * <p>
+     * This event has a {@link Result}.
+     * If the result is {@link Result#ALLOW}, the tool will be considered correct for breaking the block.
+     * If the result is {@link Result#DENY}, the tool will not be considered correct for breaking the block.
+     * If the result is {@link Result#DEFAULT}, further vanilla checks will run to determine whether the tool is considered correct for breaking the block;
+     * this is vanilla behavior.
+     */
+    public static class ToolCheck extends BlockEvent {
```
Comment (Minecraftschurli) [inline src/main/java/net/neoforged/neoforge/event/level/BlockEvent.java]: if you have a result you need to annotate with `@HasResult`

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

**Comment by Shadows-of-Fire** [comment]
If this can be made as a total replacement for HarvestCheck imo we may as well go that route.

Additionally, it should be made a standalone class instead of an inner class in BlockEvent as we move away from that pattern.

**Comment by Minecraftschurli** [comment]
if we do we should only deprecate the old event for removal

File: src/main/java/net/neoforged/neoforge/event/level/BlockEvent.java (hunk 2 above)
Comment (IchHabeHunger54) [reply src/main/java/net/neoforged/neoforge/event/level/BlockEvent.java]: HasResult is marked as deprecated, but will add.

**Comment by IchHabeHunger54** [comment]
Applied the feedback, making this a replacement for `PlayerEvent.HarvestCheck`. `PlayerEvent.HarvestCheck` and related methods have been marked as deprecated and are slated for removal in 1.20.5, to not further break compat in 1.20.4.

File: src/main/java/net/neoforged/neoforge/event/EventHooks.java [hunk 3]
```diff
@@ -181,6 +182,10 @@ public static NeighborNotifyEvent onNeighborNotify(Level level, BlockPos pos, Bl
         return event;
     }
 
+    /**
+     * DEPRECATED use {@link net.neoforged.neoforge.common.CommonHooks#isCorrectToolForDrops(BlockGetter, BlockPos, BlockState, Player)} instead
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/event/EventHooks.java]: You can use the `@deprecated` tag. Same for the other use below.

**Review by Technici4n — CHANGES_REQUESTED** [review]
Why not just add the position in the HarvestCheck event?

**Comment by Minecraftschurli** [comment]
because that is a player event and this new event is in a living entity context, I prefer the new event as changing the old one can lead to confusion and replacing seems better

**Comment by Technici4n** [comment]
Yet all the call sites pass a `Player`, and all the vanilla block breaking code uses `Player`s. I don't see the justification for extending the event to `LivingEntity`.

File: src/main/java/net/neoforged/neoforge/common/CommonHooks.java [hunk 4]
```diff
@@ -208,7 +212,14 @@ public static boolean canContinueUsing(@NotNull ItemStack from, @NotNull ItemSta
         return false;
     }
 
-    public static boolean isCorrectToolForDrops(@NotNull BlockState state, @NotNull Player player) {
+    public static boolean isCorrectToolForDrops(@NotNull BlockGetter level, @NotNull BlockPos pos, @NotNull BlockState state, @NotNull Player player) {
+        if (level instanceof LevelAccessor accessor) { //BlockEvent expects a LevelAccessor
```
Comment (Minecraftschurli) [inline src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: add a todo to change the parameter in the future too

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

**Comment by IchHabeHunger54** [comment]
Regarding the change from `Player` to `LivingEntity`, the change makes the entity not needed in the first place, which is the primary motivation. Through this event, you can go "tool -> can this block be broken". The entity context is fully optional.

This is mainly intended so mods can fire this event too if they need it, but aren't required to provide a player if they don't have one. E.g. I imagine firing it from a block breaker kind of block that requires a tool, and that can't break blocks like obsidian if an iron pickaxe is provided. Another usecase is the Minecolonies one, where the NPCs use tools and need to determine whether their tool can mine a certain block.

**Comment by Shadows-of-Fire** [comment]
Moving away from the player requirement is a good change for automation, as the current flow requires micromanaging a fake player and setting the tool in their hand, when typically the only relevant context is the tool itself.

**Comment by Technici4n** [comment]
You will always need a fake player I think, it's the simplest solution to that problem. I'd have to ask the minecolonies devs if they have a fake player. The vanilla methods already need a player so it might be better to work with that limitation.

**Comment by Thodor12** [comment]
@Technici4n We do have a Fakeplayer instance available within our citizen entities, it's just like Shadows said, we'd have to simulate said fake player it's inventory by providing a tool and what not.

A living entity would make it more generic, but if we have to pass the ItemStack in question anyway (which is all I care for), I'm not sold on Player or LivingEntity

**Comment by Shadows-of-Fire** [comment]
> You will always need a fake player I think

The LivingEntity param is nullable, which makes it optional

**Comment by IchHabeHunger54** [comment]
Yeah, the entity context can be empty if needed. There is no fake player required anymore, there isn't even any kind of entity required anymore, which is one of the major improvements of this design. What you will always need is a tool item stack, but that can be gotten trivially from `livingentity.getMainHandItem()` if needed.

TL;DR: If you don't have an entity, that's fine. Just pass in `null` for the entity param and you're done.

**Review by Minecraftschurli — DISMISSED** [review]
looks good but if technician can't be convinced I think we'll need a maintainer vote

**Comment by Technici4n** [comment]
@Minecraftschurli this PR has issues at the moment.

- The PR is arguing that we should not require a `Player` for automation reasons. Then why are all the methods still accepting a `Player` only? Notably the vanilla `playerWillDestroy`, `CommonHooks#isCorrectToolForDrops`, `IBlockExtension#canHarvestBlock`.
- Second, if a `null` entity is passed then you can't even check the currently held tool.
- Third, providing a `Player` is still required for various other checks that these automation mods should still perform (e.g. firing an event to make sure the player is allowed to break blocks, etc).

All in all, I think that providing a `FakePlayer` is the most simple and flexible solution to imitate player actions, even if it is a bit error-prone.

Finally, I will note that "block entity dependent tool requirements" as requested by @Thodor12 seem to be already available in the API by overriding `IBlockExtension#canHarvestBlock` and checking the currently held tool.

**Comment by Minecraftschurli** [comment]
@Technici4n you are always talking about checking the held tool... did you even read the event class? the new event has the tool directly so it is not needed to check the held tool. the parameters are still the same as to not be a breaking change as the old event uses the player, it would be possible to change the Player parameter on the helper method to LivingEntity and hide the old event behind an instanceof.

**Comment by Technici4n** [comment]
The event has a held tool but the block extension method does not...

**Comment by Thodor12** [comment]
Merely overriding `canHarvestBlock` was not even enough for my usecase, both because I don't have access to the original tool, nor the position which was attempted to mine.

For our usecase I have to load a tile entity which contains extra information where I have to determine the harvesting for, so I can't even use the event in it's current state on Domum Ornamentum it's side (where we have to define explicit harvest logic).

There are essentially two parts to my original issue:
 - It's not possible to create proper canHarvest logic due to some missing information in the event
 - It's not possible to fire off manual events (or somehow call a method) which does not require mocking a player and their curently held tool

**Comment by IchHabeHunger54** [comment]
After some discussion on Discord, this was instead reworked into expanding PlayerEvent.HarvestCheck to also have level and pos context. The event now also becomes result-driven; the old canHarvest() and setCanHarvest() add as wrappers for the `Result`.

File: src/main/java/net/neoforged/neoforge/event/entity/player/PlayerEvent.java [hunk 5]
```diff
@@ -50,39 +51,54 @@ public Player getEntity() {
...
 
         public boolean canHarvest() {
-            return this.success;
+            return getResult() == Result.ALLOW;
         }
 
         public void setCanHarvest(boolean success) {
-            this.success = success;
+            setResult(success ? Result.ALLOW : Result.DENY);
         }
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/event/entity/player/PlayerEvent.java]: Can you deprecate these methods for removal?

File: src/main/java/net/neoforged/neoforge/event/entity/player/PlayerEvent.java [hunk 6]
```diff
@@ -50,39 +51,54 @@ public Player getEntity() {
...
      * <br>
-     * This event does not have a result. {@link Event.HasResult}<br>
+     * This event has a {@link Result}.
+     * If the result is {@link Result#ALLOW}, the player is considered to be able to break the block.
+     * If the result is {@link Result#DENY}, the player is considered to be unable to break the block.
+     * If the result is {@link Result#DEFAULT}, further vanilla behavior will run.<br>
      * <br>
      * This event is fired on the {@link NeoForge#EVENT_BUS}.
      **/
+    @HasResult
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/event/entity/player/PlayerEvent.java]: As discussed on discord, the result changes don't really seem necessary.

File: patches/net/minecraft/world/entity/player/Player.java.patch [hunk 7]
```diff
@@ -98,20 +98,26 @@
...
+@@ -740,13 +_,19 @@
              f /= 5.0F;
          }
  
 +        f = net.neoforged.neoforge.event.EventHooks.getBreakSpeed(this, p_36282_, f, pos);
          return f;
      }
  
++    @Deprecated // Neo: use position-sensitive version below
      public boolean hasCorrectToolForDrops(BlockState p_36299_) {
```
Comment (Technici4n) [inline patches/net/minecraft/world/entity/player/Player.java.patch]: Wouldn't it make more sense to make the other method accept `nullable` parameters and forward to it?

**Review by Technici4n — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/world/entity/player/Player.java.patch (hunk 7 above)
Comment (IchHabeHunger54) [reply patches/net/minecraft/world/entity/player/Player.java.patch]: As the call chain basically goes new method -> event firing -> old method if event didn't yield a result, I don't think that's a good idea, no.

Comment (Technici4n) [reply patches/net/minecraft/world/entity/player/Player.java.patch]: I am making that suggestion exactly because I don't like that call chain. It's confusing for the deprecated method to still be called but later in the chain, IMO.

**Comment by Raycoms** [comment]
Why can't we have:     

boolean isCorrectToolForDrops(BlockState state, final ItemStack stack, BlockGetter level, BlockPos pos)


On the block itself? as an overridable method

**Comment by IchHabeHunger54** [comment]
I just saw something (maybe myself) completely fucked up the git tree. Ah well, I'll have to redo this in 1.20.5 then.

**Comment by Technici4n** [comment]
This PR still needs to be updated for 1.20.6.
